gateway/scaling/function_scaler.go

```diff
@@ -57,12 +68,31 @@ func (f *FunctionScaler) Scale(functionName string) FunctionScaleResult {
 			minReplicas = queryResponse.MinReplicas
 		}
 
-		log.Printf("[Scale] function=%s 0 => %d requested", functionName, minReplicas)
+		scaleResult := backoff(func(attempt int) error {
+			queryResponse, err := f.Config.ServiceQuery.GetReplicas(functionName)
+			if err != nil {
+				return err
+			}
+
+			f.Cache.Set(functionName, queryResponse)
+
+			if queryResponse.Replicas > 0 {
+				return nil
+			}
+
+			log.Printf("[Scale %d] function=%s 0 => %d requested", attempt, functionName, minReplicas)
```

Comment:
@LucasRoesler this is the other issue with not allowing the routine to know about the attempt - since here we can access variables within the closure that we couldn't in an other loop/function.

Reply:
ok

Reply:
Given you didn't want to see the attempt integer in the routine, how would you approach this problem?

Reply:
I would have passed a value to the backoff directly

```golang
func backoff(name string, r routine, attempts int, interval time.Duration) error {
	var err error
 	for i := 0; i < attempts; i++ {
		res := r(i)
		if res != nil {
			err = res
 			log.Printf("%s attempt: %d of %d, had error: %s\n", name, attempts, i, res)
		} else {
 			log.Printf("%s attempt: %d of %d, succeeded\n", name, attempts, i)
			err = nil
			break
		}
		time.Sleep(interval)
	}
	return err
}
```

and then, I would remove the log statement from the routine completely

```golang
routineName := fmt.Sprintf("[Scale] function=%s 0 => %d", functionName, minReplicas)
scaleResult := backoff(routineName, func() error {
...
 }
```